### PR TITLE
Updates the integration with Spring Boot with support for Root, EmbeddedStorageFoundationCustomizer and StorageManagerInitializer. Fixes #384

### LINKED ISSUE #384: Spring Boot integration: Add interceptor logic to store marked objects at end of method.
In order to align the functionality of the integrations, add interceptor logic to store marked objects at end of method.

This is similar to the functionality added in #377 

We can collect the changed objects (objects marked as dirty) using a fluent API and those are stored at the end of the method

public interface DirtyMarker {
    <T> T mark(T instance);
}
This integration allows to clearly indicate what is changed and stores the changes in an asynchronous way (synchronous is also possible through the annotation or configuration).

### PR DESCRIPTION
**Pull request by rdebusscher**

### CODE REVIEW

**Comment by rdebusscher** [comment]
Missing (will be provided shortly before PR is marked as final)
- Integration test
- Documentation

**Comment by rdebusscher** [comment]
See also readme of https://github.com/rdebusscher/microstream-spring-boot-patterns for some reasoning.

File: integrations/spring-boot/src/it/core-config/src/main/resources/application.properties [hunk 1]
```diff
@@ -7,4 +7,6 @@ one.microstream.data-file-minimum-size=1MB
 one.microstream.data-file-maximum-size=2MB
 one.microstream.data-file-minimum-use-ratio=0.9
 one.microstream.data-file-cleanup-head-file=true
-one.microstream.entity-cache-timeout=1d
\ No newline at end of file
+one.microstream.entity-cache-timeout=1d
+
+one.microstream.auto-start=false
```
Comment (zdenek-jonas) [inline integrations/spring-boot/src/it/core-config/src/main/resources/application.properties:12]: one.microstream.storage.auto-start
will be better. Microstream contains more features than storage only.

File: integrations/spring-boot/src/main/java/one/microstream/integrations/spring/boot/types/interceptor/InstanceStorer.java [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+    }
+
+    @PostConstruct
+    public void init()
+    {
+        this.signalledToStop = false;
+        this.initializePump();  // Start thread
+    }
+
+    @PreDestroy  // FIXME This doesn't seems to be working! It is never called By Spring
```
Comment (zdenek-jonas) [inline integrations/spring-boot/src/main/java/one/microstream/integrations/spring/boot/types/interceptor/InstanceStorer.java]: FIXME?

File: integrations/spring-boot/pom.xml [hunk 3]
```diff
@@ -23,6 +23,23 @@
             <version>${org.springframework.boot.version}</version>
             <scope>provided</scope>
         </dependency>
+
+        <dependency>
+            <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot-starter-web-services</artifactId>
+            <version>${org.springframework.boot.version}</version>
```
Comment (zdenek-jonas) [inline integrations/spring-boot/pom.xml]: Why we need a spring web-services in this module?

File: integrations/spring-boot/src/main/java/one/microstream/integrations/spring/boot/types/interceptor/InstanceStorer.java (hunk 2 above)
Comment (rdebusscher) [reply integrations/spring-boot/src/main/java/one/microstream/integrations/spring/boot/types/interceptor/InstanceStorer.java]: Made proper verifications and removed/replaced the comment.

File: integrations/spring-boot/pom.xml (hunk 3 above)
Comment (rdebusscher) [reply integrations/spring-boot/pom.xml]: It was a left-over from an earlier version with more functionality that required it. Removed.

File: integrations/spring-boot/src/main/java/one/microstream/integrations/spring/boot/types/config/StorageManagerFactory.java [hunk 4]
```diff
@@ -36,20 +36,34 @@
 import org.springframework.core.env.Environment;
 import org.springframework.core.env.MutablePropertySources;
 
-import one.microstream.reflect.ClassLoaderProvider;
-import one.microstream.storage.embedded.configuration.types.EmbeddedStorageConfigurationBuilder;
-import one.microstream.storage.embedded.types.EmbeddedStorageFoundation;
-import one.microstream.storage.embedded.types.EmbeddedStorageManager;
-import one.microstream.util.logging.Logging;
+import java.util.*;
```
Comment (zdenek-jonas) [inline integrations/spring-boot/src/main/java/one/microstream/integrations/spring/boot/types/config/StorageManagerFactory.java]: This is not used in Microstream, please remove and use full imports.

File: integrations/spring-boot/src/main/java/one/microstream/integrations/spring/boot/types/config/StorageManagerFactory.java [hunk 5]
```diff
@@ -36,20 +36,34 @@
...
 
 @Configuration
-@EnableConfigurationProperties(MicrostreamConfigurationProperties.class)
-public class MicrostreamConfiguration
+public class StorageManagerFactory
 {
 
     private static final String PREFIX = "one.microstream.";
-    Logger logger = Logging.getLogger(MicrostreamConfiguration.class);
+    Logger logger = Logging.getLogger(StorageManagerFactory.class);
```
Comment (zdenek-jonas) [inline integrations/spring-boot/src/main/java/one/microstream/integrations/spring/boot/types/config/StorageManagerFactory.java]: private final static?

File: integrations/spring-boot/src/main/java/one/microstream/integrations/spring/boot/types/storage/StorageBeanFactory.java [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+        for (String definitionName : beanDefinitionRegistry.getBeanDefinitionNames())
+        {
+
+            final BeanDefinition beanDefinition = beanDefinitionRegistry.getBeanDefinition(definitionName);
+            if (beanDefinition.getBeanClassName() != null)
+            {
+                try
+                {
+                    // Don't use generics as that will not work with the Supplier that provides the Root object further in this method
+                    final Class clazz = Class.forName(beanDefinition.getBeanClassName());
```
Comment (zdenek-jonas) [inline integrations/spring-boot/src/main/java/one/microstream/integrations/spring/boot/types/storage/StorageBeanFactory.java:107]: image
There are warnings here, is it ok?

File: integrations/spring-boot/src/main/java/one/microstream/integrations/spring/boot/types/DirtyMarker.java [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+ *     ...
+ *     dirtyMarker.mark(product).setName(newName);
+ * }
+ * </pre>
+ */
+public interface DirtyMarker
+{
+
+    <T> T mark(T instance);
+}
```
Comment (zdenek-jonas) [inline integrations/spring-boot/src/main/java/one/microstream/integrations/spring/boot/types/DirtyMarker.java]: Should this class remain here?

**Review by zdenek-jonas — COMMENTED** [review]
There are some small questions. But generally ok.

File: docs/modules/misc/pages/integrations/spring-boot.adoc [hunk 8]
```diff
@@ -8,14 +8,114 @@ It is available within this artifact:
...
+
+The configuration of the _StorageManager_ can be done using key/value pairs that are provided by Spring Boot external Configuration.  The configuration keys must be prefixed by `one.microstream`
+
+[source]
+----
+one.microstream.storage-directory=/opt/data/microstream
+one.microstream.channel-count=2
+----
+
+The list of all MicroStream configuration properties and their meaning are listed on our https://docs.microstream.one/manual/storage/configuration/properties.html[documentation page].
```
Comment (fh-ms) [inline docs/modules/misc/pages/integrations/spring-boot.adoc]: No need to use the external URL for internal doc links. You can use [xrefs](https://docs.antora.org/antora/2.2/asciidoc/page-to-page-xref/) for that:
`xref:storage:configuration/properties.adoc[documentation page]`

File: integrations/spring-boot/src/main/java/one/microstream/integrations/spring/boot/types/config/StorageManagerFactory.java (hunk 5 above)
Comment (fh-ms) [reply integrations/spring-boot/src/main/java/one/microstream/integrations/spring/boot/types/config/StorageManagerFactory.java:48]: same same, we are not too picky here

**Review by fh-ms — CHANGES_REQUESTED** [review]

**Review by fh-ms — APPROVED** [review]

**Review by zdenek-jonas — APPROVED** [review]

**Pull request merged**